### PR TITLE
[metadata] Move MonoClass initialization to a separate file

### PR DESCRIPTION
**Pull request by lambdageek**

### CODE REVIEW

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -4721,7 +4721,7 @@ add_generic_class_with_depth (MonoAotCompile *acfg, MonoClass *klass, int depth,
 	if (klass->image == mono_defaults.corlib && !strcmp (klass->name_space, "System.Collections.Generic") &&
 		(!strcmp(klass->name, "ICollection`1") || !strcmp (klass->name, "IEnumerable`1") || !strcmp (klass->name, "IList`1") || !strcmp (klass->name, "IEnumerator`1") || !strcmp (klass->name, "IReadOnlyList`1"))) {
 		MonoClass *tclass = mono_class_from_mono_type (mono_class_get_generic_class (klass)->context.class_inst->type_argv [0]);
-		MonoClass *array_class = mono_bounded_array_class_get (tclass, 1, FALSE);
+		MonoClass *array_class = mono_bounded_array_class_create (tclass, 1, FALSE);
```
Comment (jaykrell) [inline mono/mini/aot-compiler.c]: Someone will object to name lengthening?

Comment (jaykrell) [reply mono/mini/aot-compiler.c]: But it is probably good -- get tends to retrieve available state, create tends to create new/additional state.

**Comment by vargaz** [comment]
Isn't the whole class.c file about writing to MonoClass ?

File: mono/mini/aot-compiler.c (hunk 1 above)
Comment (lambdageek) [reply mono/mini/aot-compiler.c]: The main difference is that `mono_bounded_array_get` is `MONO_API`, so fixed in stone.  `mono_bounded_array_class_create` can have its semantics change: for example to be relaxed to create array classes that are not fully initialized.

**Comment by lambdageek** [comment]
> Isn't the whole class.c file about writing to MonoClass ?

I don't think so - there are several algorithms in there: is assignable from, for example; also searches of various kinds - find field by name, find method's defining class, etc.

Actually if it turns out that it was 30% getters and 70% MonoClass initialization, I'll drop this PR and just move the 30% out of the way.

**Comment by lambdageek** [comment]
Well... if nothing else... `class.c` goes from 340 KiB to 175KiB and 165KiB in `class-init.c`.

I think the blame is mostly preserved (if you use `git blame -C mono/metadata/class-init.c` on the command line - GitHub doesn't seem to track moves across files)

**Comment by vargaz** [comment]
We might want to move vtable creation to a separate file too, its 1k lines and fairly self contained.

**Comment by luhenry** [comment]
@vargaz that's a good idea, let's do it after we are done with `MonoClass` initialization refactoring.

File: mono/metadata/class-init.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+MONO_BEGIN_DECLS
+
+void
+mono_classes_init (void);
+
+void
+mono_classes_cleanup (void);
+
+MonoClass *
+mono_class_create_from_typedef (MonoImage *image, guint32 type_token, MonoError *error);
```
Comment (luhenry) [inline mono/metadata/class-init.h]: Could we rename all the functions in this header to something like `mono_class_{init,create}_*`. For example they could be in order:
 - mono_class_create_from_typedef
 - mono_class_create_from_generic_class
 - mono_class_create_bounded_array
 - mono_class_create_array
 - mono_class_create_from_generic_parameter_internal
 - mono_class_create_ptr
 - mono_class_create_fnptr
 - mono_class_init_vtable_general
 - mono_class_init_sizes
 - mono_class_init_basic_field_info
 - ...

Comment (luhenry) [reply mono/metadata/class-init.h]: And for the functions that are MONO_API, we can just keep their old version in `class.c` and have them call the new version in `class-init.c`.

Comment (lambdageek) [reply mono/metadata/class-init.h]: Sure. Normalizing the names is a great idea.  I would prefer to do keep this PR focused on just moving functions to a separate file as much as possible, and do a followup for renaming.

Comment (lambdageek) [reply mono/metadata/class-init.h]: Oh, but a few of these are new in this PR. I'll update those to use the `mono_class_create_foo`/`mono_class_init_foo` naming.

Comment (lambdageek) [reply mono/metadata/class-init.h]: Ok, renamed all the creation functions to be `mono_class_create_XXX`.  The init/setup ones I think I still want to do in a separate PR, I think.

**Comment by lambdageek** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by lambdageek** [comment]
@monojenkins build Linux AArch64

**Comment by lambdageek** [comment]
The "Linux AArch64 Interpreter" failure doesn't seem related (I see it on other PRs too - #6685 for example):

```
test-404.cs... REGRESSION (SUCCESS -> EXECUTION ERROR)
```

**Comment by lambdageek** [comment]
@monojenkins build

**Comment by luhenry** [comment]
It looks all good, but let's just merge it after we branch `2018-02` on monday.

**Pull request merged**